File: dotnet/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/SpeedBoostDecorator.cs
namespace DecoratorDesignPattern.Example1;

public class SpeedBoostDecorator : CharacterDecorator
{
    private readonly string _boostName;
    private readonly int _speedBonus;

    public SpeedBoostDecorator(ICharacter character, string boostName, int speedBonus)
        : base(character)
    {
        if (speedBonus <= 0)
            throw new ArgumentException("Hız bonusu pozitif olmalıdır.");

        _boostName = boostName ?? "Hız Artırıcı";
        _speedBonus = speedBonus;
    }

    public override int GetSpeed()
    {
        return base.GetSpeed() + _speedBonus;
    }

    public override string GetDescription()
    {
        return $"{base.GetDescription()} + {_boostName} (+{_speedBonus} Hız)";
    }
}

File: nodejs/DesignPatterns/StructuralDesignPatterns/DecoratorDesignPattern/DecoratorDesignPattern.Example1/src/SpeedBoostDecorator.ts
import { ICharacter } from './ICharacter';
import { CharacterDecorator } from './CharacterDecorator';

export class SpeedBoostDecorator extends CharacterDecorator {
  constructor(
    character: ICharacter,
    private readonly boostName: string,
    private readonly speedBonus: number,
  ) {
    super(character);
    if (speedBonus <= 0) throw new Error('Hız bonusu pozitif olmalıdır.');
  }

  getSpeed(): number { return super.getSpeed() + this.speedBonus; }
  getDescription(): string { return `${super.getDescription()} + ${this.boostName} (+${this.speedBonus} Hız)`; }
}
